File: java_spring-boot_spring-web-mvc_spring-data-jpa_lombok/src/main/java/org/example/demoapp/company/CompanyService.java
package org.example.demoapp.company;

import javax.transaction.Transactional;

import java.util.List;
import java.util.Optional;

import org.example.demoapp.employee.Employee;
import org.example.demoapp.employee.EmployeeNotFoundException;
import org.example.demoapp.employee.EmployeeService;
import org.springframework.stereotype.Service;

@Service
@Transactional
public class CompanyService {

    private final CompanyRepository companyRepository;
    private final EmployeeService employeeService;

    public CompanyService(CompanyRepository companyRepository,
                          EmployeeService employeeService) {
        this.companyRepository = companyRepository;
        this.employeeService = employeeService;
    }

    public List<Company> findAll() {
        return companyRepository.findAll();
    }

    public Optional<Company> findById(long id) {
        return companyRepository.findById(id);
    }

    public Company save(Company company) {
        return companyRepository.save(company);
    }

    public void delete(long id) {
        companyRepository.findById(id)
                .ifPresent(companyRepository::delete);
    }

    public void addEmployee(long companyId, long employeeId) {
        Company company = companyRepository.findById(companyId)
                .orElseThrow(CompanyNotFoundException::new);

        if (company.isEmployeeEmployed(employeeId)) {
            throw new EmployeeAlreadyEmployedInCompanyException(companyId, employeeId);
        }

        Employee employee = employeeService.findById(employeeId)
                .orElseThrow(EmployeeNotFoundException::new);

        company.addEmployee(employee);
    }

    public void removeEmployee(long companyId, long employeeId) {
        companyRepository.findById(companyId)
                .ifPresent(company -> company.removeEmployee(employeeId));
    }
}

File: kotlin_spring-boot_spring-boot-data-jpa_spring-web_maven/src/main/kotlin/org/example/demoapp/company/CompanyService.kt
package org.example.demoapp.company

import org.example.demoapp.employee.EmployeeNotFoundException
import org.example.demoapp.employee.EmployeeService
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Service
@Transactional
class CompanyService(val companyRepository: CompanyRepository,
                     val employeeService: EmployeeService) {

    fun findAll(): List<Company> {
        return companyRepository.findAll()
    }

    fun findById(id: Long): Optional<Company> {
        return companyRepository.findById(id)
    }

    fun save(company: Company): Company {
        return companyRepository.save(company)
    }

    fun delete(id: Long) {
        companyRepository.findById(id)
                .ifPresent { company: Company -> companyRepository.delete(company) }
    }

    fun addEmployee(companyId: Long, employeeId: Long) {
        val company = companyRepository.findById(companyId)
                .orElseThrow { CompanyNotFoundException() }

        if (company.alreadeEmployed(employeeId)) {
            throw EmployeeAlreadyEmployedInCompanyException()
        }

        val employee = employeeService.findById(employeeId)
                .orElseThrow { EmployeeNotFoundException() }

        company.employees.add(employee)
    }

    fun removeEmployee(companyId: Long, employeeId: Long) {
        companyRepository.findById(companyId).ifPresent { company ->
            company.employees.removeIf { employee -> employee.id == employeeId }
        }
    }
}